lib/redis_mutex.ex

```diff
@@ -1,90 +1,76 @@
 defmodule RedisMutex do
   @moduledoc """
-  An Elixir library for using Redis locks
-
-  ## Setup
+  An Elixir library for using Redis locks.
+  """
 
+  @type connection_options :: [
+          host: String.t(),
+          port: non_neg_integer(),
+          database: String.t() | non_neg_integer(),
+          username: String.t(),
+          password: Redix.password(),
+          timeout: timeout(),
+          sync_connect: boolean(),
+          exit_on_disconnection: boolean(),
+          backoff_initial: non_neg_integer(),
+          backoff_max: timeout(),
+          ssl: boolean(),
+          socket_opts: list(term()),
+          hibernate_after: non_neg_integer(),
+          spawn_opt: keyword(),
+          debug: keyword(),
+          sentinel: keyword()
+        ]
 
-  1. Add `redis_mutex` to your list of dependencies in `mix.exs`:
+  @type start_options :: {:redis_url, String.t()} | connection_options()
 
-    ```elixir
-    def deps do
-      [{:redis_mutex, "~> 0.4.0"}]
-    end
-    ```
+  @default_lock_module RedisMutex.Lock
 
-  2. Ensure `redis_mutex` is started before your application:
+  def child_spec(opts \\ []) do
+    args = child_spec_args(opts)
 
-    ```elixir
-    def application do
-      [applications: [:redis_mutex]]
-    end
-    ```
+    %{
+      id: __MODULE__,
+      start: {__MODULE__, :start_link, [args]},
+      type: :supervisor
+    }
+  end
 
-  3. Set the `redis_url` in your `config.exs`
+  @spec start_link(Keyword.t()) :: Supervisor.on_start()
+  def start_link(opts) do
+    lock_module = lock_module(opts)
 
-    ```elixir
-    config :redis_mutex, redis_url: {:system, "REDIS_URL"}
-    ```
+    options = Application.get_env(:redis_mutex, :redis_options, [])
 
-  4. Call `use RedisMutex` in the module you want to use the lock.
+    RedisMutex.Supervisor.start_link(
+      lock_module,
+      options
+    )
+  end
 
-    ```elixir
-    defmodule PossumLodge do
+  defmacro __using__(opts \\ []) do
+    lock_module = lock_module(opts)
 
-      def get_oath do
-        "Quando omni flunkus moritati"
-      end
+    quote do
+      import unquote(lock_module), warn: false
     end
-    ```
-
-    With a Redis lock:
-
-    ```elixir
-    defmodule PossumLodge do
-      use RedisMutex
+  end
```

Comment:
From the Elixir docs on [use/2](https://hexdocs.pm/elixir/Kernel.html#use/2)

> If you can, avoid use in favor of [import/2](https://hexdocs.pm/elixir/Kernel.SpecialForms.html#import/2) or [alias/2](https://hexdocs.pm/elixir/Kernel.SpecialForms.html#alias/2) whenever possible.

Should modules just be importing the lock module?

Reply:
> Should modules just be importing the lock module?

Doing `use RedisMutex` allows us to use different lock modules. One goal of being able to set a different lock module in test is so that people can define a double for the lock and run unit tests without having Redis up. If we wanted to import the lock module, what mechanism would work for allowing someone to run unit tests without Redis up?

Reply:
you could use config instead and Application.compile_env?

Reply:
This section has been replaced